File: atividade-25/atividade.py
class AFN:
    def __init__(self):
        self.states = {'q0', 'q1', 'q2'}
        self.start_state = 'q0'
        self.accept_states = {'q1', 'q2'}

    def is_accepted(self, string):
        current_states = {self.start_state}  

        for char in string:
            next_states = set()
            for state in current_states:
                if state == 'q0':  
                    if char == 'a':
                        next_states.add('q0')  
                    elif char == 'b':
                        next_states.add('q1')  
                elif state == 'q1':  
                    if char == 'b':
                        next_states.add('q1')  

            current_states = next_states

        
        return 'q1' in current_states or 'q0' in current_states

if __name__ == '__main__':
    afn = AFN()
    print(afn.is_accepted("aaa"))    
    print(afn.is_accepted("aabbb"))  
    print(afn.is_accepted("aaab"))   
    print(afn.is_accepted("aabb"))   
    print(afn.is_accepted("abba"))   
    print(afn.is_accepted("b"))      
    print(afn.is_accepted("ba"))     

File: atividade-25/atividade.go
package main

import "fmt"

type AFN struct {
	states       map[string]bool
	startState   string
	acceptStates map[string]bool
}

func NewAFN() *AFN {
	return &AFN{
		states: map[string]bool{
			"q0": true,
			"q1": true,
		},
		startState: "q0",
		acceptStates: map[string]bool{
			"q0": true,
			"q1": true,
		},
	}
}

func (afn *AFN) isAccepted(input string) bool {
	currentStates := map[string]bool{
		afn.startState: true,
	}

	for _, c := range input {
		nextStates := make(map[string]bool)
		for state := range currentStates {
			switch state {
			case "q0":
				if c == 'a' {
					nextStates["q0"] = true
				} else if c == 'b' {
					nextStates["q1"] = true
				}
			case "q1":
				if c == 'b' {
					nextStates["q1"] = true
				}
			}
		}
		currentStates = nextStates
	}

	_, acceptedInQ0 := currentStates["q0"]
	_, acceptedInQ1 := currentStates["q1"]
	return acceptedInQ0 || acceptedInQ1
}

func main() {
	afn := NewAFN()
	fmt.Println(afn.isAccepted("aaa"))   
	fmt.Println(afn.isAccepted("aabbb")) 
	fmt.Println(afn.isAccepted("aaab"))  
	fmt.Println(afn.isAccepted("aabb"))  
	fmt.Println(afn.isAccepted("abba"))  
	fmt.Println(afn.isAccepted("b"))     
	fmt.Println(afn.isAccepted("ba"))    
}
